In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
creditDF = pd.read_csv("lending_data.csv")

# Review the DataFrame
creditDF.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = creditDF["loan_status"]

# Separate the X variable, the features
X = creditDF.drop(columns="loan_status")

In [4]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train, x_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logClassifier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using training data
logClassifier.fit(x_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
predictedValues = logClassifier.predict(x_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced accuracy score of the model
balancedAccuracyScore = balanced_accuracy_score(y_test, predictedValues)
print(f"Balanced Accuracy Score: {(balancedAccuracyScore*100):.2f}%")

# Print the accuracy score of the model
from sklearn.metrics import accuracy_score

accuracyScore = accuracy_score(y_test, predictedValues)
print(f"Accuracy Score: {(accuracyScore*100):.2f}%")

Balanced Accuracy Score: 95.20%
Accuracy Score: 99.18%


In [16]:
# Generate a confusion matrix for the model
print(confusion_matrix(y_test, predictedValues))

[[18663   102]
 [   56   563]]


In [33]:
# confusion matrix as a dataframe
confusionMatrixData = confusion_matrix(y_test, predictedValues)
confusionMatrixDF = pd.DataFrame(
    confusionMatrixData,
    index=["Outcome is 0 (Healthy Loan)", "Outcome is 1 (High Risk Loan)"],
    columns=["Test says 0 (Healthy)", "Test says 1 (High Risk Loan)"]
)
confusionMatrixDF

Test says 0 (Healthy)  \
Outcome is 0 (Healthy Loan)                    18663   
Outcome is 1 (High Risk Loan)                     56   

                               Test says 1 (High Risk Loan)  
Outcome is 0 (Healthy Loan)                             102  
Outcome is 1 (High Risk Loan)                           563

In [18]:
# Print the classification report for the model
print(classification_report(y_test, predictedValues))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



In [50]:
trueHealthy = confusionMatrixData[0,0]
falseHighRisk = confusionMatrixData[0,1]
trueHighRisk = confusionMatrixData[1,1]
falseHealthy = confusionMatrixData[1,0]

print(f"The true healthy count is: {trueHealthy}")
print(f"The true high risk count is: {trueHighRisk}")
print(f"The false healthy count is: {falseHealthy}")
print(f"The false high risk count is: {falseHighRisk}")
print("------------------------------------------------")
print(f"The accuracy score is: {((trueHealthy+trueHighRisk)/(trueHealthy + falseHealthy + trueHighRisk + falseHighRisk))*100:.3f}%")
print(f"The precision of healthy loans is {(trueHealthy/(trueHealthy+ falseHealthy))*100:.3f}% (0 Precision score)")
print(f"The precision of high risk loans is {(trueHighRisk/(trueHighRisk+ falseHighRisk))*100:.3f}% (1 Precision score)")

The true healthy count is: 18663
The true high risk count is: 563
The false healthy count is: 56
The false high risk count is: 102
------------------------------------------------
The accuracy score is: 99.185%
The precision of healthy loans is 99.701% (0 Precision score)
The precision of high risk loans is 84.662% (1 Precision score)


### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the cell above, the model has a 99.185% accuracy score. It also has a 99.701% precision of healthy loans and a 84.662% precision of high risk loans. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [21]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros_model.fit_resample(x_train, y_train)

In [23]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    56271
1    56271
Name: loan_status, dtype: int64

In [26]:
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

In [24]:
X_resampled.value_counts()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  derogatory_marks  total_debt
18500.0    10.976         83900            0.642431        12               2                 53900         209
17500.0    10.552         79900            0.624531        11               2                 49900         181
17400.0    10.510         79500            0.622642        11               2                 49500         157
18400.0    10.939         83500            0.640719        12               2                 53500         153
19300.0    11.329         87200            0.655963        12               2                 57200         140
                                                                                                           ... 
12900.0    8.594          61500            0.512195        7                1                 31500           1
           8.598          61500            0.512195        7                1                 31500           1

In [25]:
X.value_counts()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  derogatory_marks  total_debt
9600.0     7.211          48500            0.381443        4                0                 18500         78
           7.207          48400            0.380165        4                0                 18400         76
9200.0     7.018          46600            0.356223        3                0                 16600         75
9600.0     7.204          48400            0.380165        4                0                 18400         75
9400.0     7.102          47400            0.367089        3                0                 17400         73
                                                                                                            ..
16700.0    10.238         76900            0.609883        10               2                 46900          1
           10.236         76900            0.609883        10               2                 46900          1
       

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [27]:
# Split the data using train_test_split
# Assign a random_state of 1 to the function
x_train1, x_test1, y_train1, y_test1 = train_test_split(X_resampled, y_resampled, random_state=1)

In [28]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logClassifier1 = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
logClassifier1.fit(x_train1, y_train1)

# Make a prediction using the testing data
predictedValues1 = logClassifier1.predict(x_test1)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [29]:
# Print the balanced accuracy score of the model
balancedAccuracyScore1 = balanced_accuracy_score(y_test1, predictedValues1)
print(f"Balanced Accuracy Score (Resampled): {(balancedAccuracyScore1*100):.2f}%")

accuracyScore1 = accuracy_score(y_test1, predictedValues1)
print(f"Accuracy Score (Resampled): {(accuracyScore1*100):.2f}%")

Balanced Accuracy Score (Resampled): 99.42%
Accuracy Score (Resampled): 99.42%


In [30]:
# Generate a confusion matrix for the model
print(confusion_matrix(y_test1, predictedValues1))

[[14018    88]
 [   74 13956]]


In [34]:
# confusion matrix as a dataframe
confusionMatrixData1 = confusion_matrix(y_test1, predictedValues1)
confusionMatrixDF1 = pd.DataFrame(
    confusionMatrixData1,
    index=["Outcome is 0 (Healthy Loan)", "Outcome is 1 (High Risk Loan)"],
    columns=["Test says 0 (Healthy)", "Test says 1 (High Risk Loan)"]
)
confusionMatrixDF1

Test says 0 (Healthy)  \
Outcome is 0 (Healthy Loan)                    14018   
Outcome is 1 (High Risk Loan)                     74   

                               Test says 1 (High Risk Loan)  
Outcome is 0 (Healthy Loan)                              88  
Outcome is 1 (High Risk Loan)                         13956

In [32]:
# Print the classification report for the model
print(classification_report(y_test1, predictedValues1))

              precision    recall  f1-score   support

           0       0.99      0.99      0.99     14106
           1       0.99      0.99      0.99     14030

    accuracy                           0.99     28136
   macro avg       0.99      0.99      0.99     28136
weighted avg       0.99      0.99      0.99     28136



In [51]:
trueHealthy1 = confusionMatrixData1[0,0]
falseHighRisk1 = confusionMatrixData1[0,1]
trueHighRisk1 = confusionMatrixData1[1,1]
falseHealthy1 = confusionMatrixData1[1,0]

print(f"The true healthy count is: {trueHealthy1}")
print(f"The true high risk count is: {trueHighRisk1}")
print(f"The false healthy count is: {falseHealthy1}")
print(f"The false high risk count is: {falseHighRisk1}")
print("------------------------------------------------")
print(f"The accuracy score is: {((trueHealthy1+trueHighRisk1)/(trueHealthy1 + falseHealthy1 + trueHighRisk1 + falseHighRisk1))*100:.3f}%")
print(f"The precision of healthy loans is {(trueHealthy1/(trueHealthy1+ falseHealthy1))*100:.3f}% (0 Precision score)")
print(f"The precision of high risk loans is {(trueHighRisk1/(trueHighRisk1+ falseHighRisk1))*100:.3f}% (1 Precision score)")

The true healthy count is: 14018
The true high risk count is: 13956
The false healthy count is: 74
The false high risk count is: 88
------------------------------------------------
The accuracy score is: 99.424%
The precision of healthy loans is 99.475% (0 Precision score)
The precision of high risk loans is 99.373% (1 Precision score)


### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** According to the cell above, the model has a 99.424% accuracy score. It also has a 99.475% precision of healthy loans and a 99.373% precision of high risk loans.

**SUMMARY:** The resampled Training Data vs Training Data Difference Values

- Accuracy: +0.239%
- Precision (healthy loans): -0.226%
- Precision (high risk loans): +14.711%

Due to the precision difference value for high risk loans, the resampled training data is my recommended model.